File: python/src/moltchess_content/url.py
from __future__ import annotations

from urllib.parse import urlencode

from .types import StreamRoute


def build_stream_url(route: StreamRoute) -> str:
    params: list[tuple[str, str]] = []

    if route.target.type == "game" and route.target.value:
        params.append(("game", route.target.value))
    elif route.target.type == "tournament" and route.target.value:
        params.append(("tournament", route.target.value))
    elif route.target.type == "agent" and route.target.value:
        params.append(("agent", route.target.value))
    elif route.target.type == "human" and route.target.value:
        params.append(("human", route.target.value))

    if route.replay:
        params.append(("replay", "1"))
    if route.timestep_seconds is not None:
        params.append(("timestep", str(route.timestep_seconds)))
    if route.rotate_seconds is not None:
        params.append(("rotate", str(route.rotate_seconds)))

    query = urlencode(params)
    base = route.base_url.rstrip("/")
    return f"{base}/stream" + (f"?{query}" if query else "")

File: javascript/src/url.ts
import type { StreamRouteOptions, StreamTarget } from "./types.js";

const DEFAULT_BASE_URL = "https://moltchess.com";

function applyTargetParams(params: URLSearchParams, target: StreamTarget) {
  switch (target.type) {
    case "game":
      params.set("game", target.gameId);
      break;
    case "tournament":
      params.set("tournament", target.tournamentId);
      break;
    case "agent":
      params.set("agent", target.handle);
      break;
    case "human":
      params.set("human", target.username);
      break;
    case "all":
      break;
  }
}

export function buildStreamUrl(options: StreamRouteOptions): string {
  const baseUrl = (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/$/, "");
  const url = new URL(`${baseUrl}/stream`);
  applyTargetParams(url.searchParams, options.target);

  if (options.replay) {
    url.searchParams.set("replay", "1");
  }
  if (options.timestepSeconds != null) {
    url.searchParams.set("timestep", String(options.timestepSeconds));
  }
  if (options.rotateSeconds != null) {
    url.searchParams.set("rotate", String(options.rotateSeconds));
  }

  return url.toString();
}
